File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-typescript/src/records.ts
// Copyright (C) 2026 Antony Stubbs and contributors

/**
 * What a processor is handed, and what it may ask to be produced.
 *
 * Keys and values are `Buffer` - the bytes Kafka held. The proxy never deserializes, so
 * deserialization is the application's, in the application's own types. `null` is not `Buffer[0]`:
 * Kafka distinguishes a null key or a tombstone value from an empty one, and so does this surface.
 */

/**
 * One record as delivered, carrying the Parallel-Consumer-derived delivery state an in-process user
 * function would see. Nothing transport-specific appears here - no epoch, no token, no connection
 * identity - because the fencing token is the client's to echo and never the application's to read.
 */
export interface InboundRecord {
  readonly topic: string;
  readonly partition: number;
  /**
   * `bigint`, because a Kafka offset is a 64-bit integer and `number` silently loses precision
   * past 2^53. Compare with `0n`, not `0`.
   */
  readonly offset: bigint;
  /** The key bytes, or `null` for a keyless record. */
  readonly key: Buffer | null;
  /** The value bytes, or `null` for a tombstone - which is not the same as an empty value. */
  readonly value: Buffer | null;
  /** 1 on first delivery, 2 on the first redelivery. */
  readonly attempt: number;
  /** When the previous attempt failed; `null` before the first failure. */
  readonly lastFailureAt: Date | null;
  /**
   * The reason recorded for the previous failed attempt; `null` before the first failure. Worker
   * supplied and may embed record payload: treat as untrusted input.
   */
  readonly lastFailureReason: string | null;
}

/**
 * A record a successful outcome asks Parallel Consumer to produce - the only sanctioned route for a
 * processor's Kafka output. Workers never produce directly; the proxy produces with its own
 * producer before the input record's offset may become eligible to commit.
 */
export interface OutboundRecord {
  readonly topic: string;
  /** The key bytes to produce, or `null` for a keyless record. */
  readonly key?: Buffer | null;
  /** The value bytes to produce, or `null` for a tombstone. */
  readonly value?: Buffer | null;
}

File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-python/src/parallel_consumer/records.py
# Copyright (C) 2026 Antony Stubbs and contributors

"""The records that cross the user's function: one in, zero or more out.

Keys and values are ``bytes``. This library never deserializes - the proxy does not either, and
serialization formats are the user's business, in the user's own code.
"""

from __future__ import annotations

import dataclasses
from datetime import datetime

__all__ = ["InboundRecord", "OutboundRecord"]


@dataclasses.dataclass(frozen=True)
class InboundRecord:
    """One record as delivered to the user's function.

    Carries the Kafka record *plus* the delivery state an in-process Parallel Consumer user
    function would see: which attempt this is, and why the previous one failed. That state is
    product data - nothing about the transport (delivery tokens, connection identity) appears
    here, deliberately.

    ``key`` and ``value`` are ``None`` when Kafka held null, which is a different thing from
    empty bytes: a tombstone has a null value.
    """

    topic: str
    partition: int
    offset: int
    key: bytes | None
    value: bytes | None
    attempt: int
    """1 on first delivery, 2 on the first redelivery."""

    last_failure_at: datetime | None = None
    """When the previous attempt failed; ``None`` before the first failure."""

    last_failure_reason: str | None = None
    """The previous attempt's reason. Worker-supplied text: untrusted, and it may quote payload."""

    def __repr__(self) -> str:
        # Deliberately omits key and value: payloads are untrusted input and do not belong in
        # log lines or tracebacks.
        return f"InboundRecord({self.topic}-{self.partition}@{self.offset}, attempt {self.attempt})"


@dataclasses.dataclass(frozen=True)
class OutboundRecord:
    """A record a successful outcome asks Parallel Consumer to produce.

    The only sanctioned route for a worker's Kafka output: workers never hold a producer, and
    the proxy produces these with its own before the input record's offset can be committed.
    """

    topic: str
    key: bytes | None = None
    value: bytes | None = None

    def __post_init__(self) -> None:
        if not self.topic:
            raise ValueError("an OutboundRecord needs a destination topic")

    def __repr__(self) -> str:
        # As above: no payload in a repr.
        return f"OutboundRecord({self.topic})"
